### PR TITLE
Chapter4/homework/1

### PR DESCRIPTION
**Pull request by mkasarci**
I've created minimum amount of unit tests that covers all the methods needs to test. But I couldn't find anywhere to use mock. Mocking data is still confusing for me. I don't understand when should I use it. I feel like my test methods has dependencies cause I didn't mock the classes that I should.

### CODE REVIEW

**Comment by Almantask** [comment]
Mocking is not needed for this hw.

File: Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        [Theory]
+        [MemberData(nameof(GetCardAndName))]
+        public void ToString_Returns_Expected_String(Card card, string expectedName)
+        {
+            var Name = card.ToString();
+
+            Assert.Equal(expectedName, Name);
+        }
+
+        public static IEnumerable<object[]> GetCardAndName()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs:20]: Don't have to be a method, can be a property. `GetCardAndName()` should be `CardNameExpectations =>`

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        [Fact]
+        public void Deck_DrawAt_Index_More_Than_Card_Count_Throws_OutOfCardsException()
+        {
+            var deck = CreateDeckWithSomeCards();
+
+             void DrawOutOfIndex() => deck.DrawAt(4);
+
+            Assert.Throws<OutOfCardsException>(DrawOutOfIndex);
+        }
+        private IDeck CreateDeck(IEnumerable<Card> cards) => new Deck(cards);
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:55]: Not needed. You're just hiding your deck creation, which is a simple as new Deck(cards);

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        private List<Card> _someCards = new List<Card>
+        {
+            new Card(Card.Suites.Clubs, Card.Ranks.Ace),
+            new Card(Card.Suites.Clubs, Card.Ranks.Two),
+            new Card(Card.Suites.Clubs, Card.Ranks.Three)
+        };
+
+        [Fact]
+        public void Deck_Given_Null_Cards_Throws_ArgumentNullException()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:18]: That's not what you are testing.
`New_Deck_Given_...`

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        public void Deck_DrawAt_Index_More_Than_Card_Count_Throws_OutOfCardsException()
+        {
+            var deck = CreateDeckWithSomeCards();
+
+             void DrawOutOfIndex() => deck.DrawAt(4);
+
+            Assert.Throws<OutOfCardsException>(DrawOutOfIndex);
+        }
+        private IDeck CreateDeck(IEnumerable<Card> cards) => new Deck(cards);
+        private IDeck CreateDeckWithSomeCards() => CreateDeck(_someCards);
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:56]: You should specify the name. In either way, avoid using shared fixtures like this one.
Keep tests minimalistic. Although 3 cards is indeed the minimalistic case, for the tests that you have now it makes no difference.

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            var deck = CreateDeckWithSomeCards();
+            Card selectedCard = deck.DrawAt(1);
+            Card expectedCard = _someCards[1];
+
+            Assert.Equal(expectedCard, selectedCard);
+        }
+
+        [Fact]
+        public void Deck_DrawFromTop_Should_Return_Last_Card()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:36]: `should` can be removed

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        [Fact]
+        public void Deck_Given_Null_Cards_Throws_ArgumentNullException()
+        {
+            void NullDeck() => CreateDeck(null);
+
+            Assert.Throws<ArgumentNullException>(NullDeck);
+        }
+
+        [Fact]
+        public void Deck_DrawAt_Returns_Expected_Card()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:26]: Doesn't explain well enough what is tested. `Deck` is not needed, it's DeckTests after all.

Comment (Almantask) [reply Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:26]: DrawAt_Given_2Cards_When_Index1_Returns_2ndCard.
However, you should test edge cases: 1st, middle, last card:
```
[Theory]
[...]
DrawAt_Given_3Cards_Returns_Expected()
```

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            var deck = CreateDeckWithSomeCards();
+
+            Card expecterdCard = _someCards[_someCards.Count - 1];
+            Card selectedCardFromTop = deck.DrawFromTop();
+
+            Assert.Equal(expecterdCard, selectedCardFromTop);
+        }
+
+        [Fact]
+        public void Deck_DrawAt_Index_More_Than_Card_Count_Throws_OutOfCardsException()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:47]: Should be:
`DrawAt_Index1_Given_NoCards_Throws_..`

Be minimalistic. You don't need 3 cards in the deck to check what happens if index is out of range. No cards and index 1 is the minimalistic scenario.

File: Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+using System.Text;
+using BootCamp.Chapter.Gambling;
+using Moq;
+using Xunit;
+
+namespace BootCamp.Chapter.Tests.GameComponents
+{
+    public class HandTests
+    {
+        private Card[] _someCards = new[]
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:14]: if you need it to be a list, then make it a list instead of array. No point in converting later.

File: Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            new Card(Card.Suites.Spades, Card.Ranks.Ace),
+            new Card(Card.Suites.Clubs, Card.Ranks.Two),
+            new Card(Card.Suites.Hearts, Card.Ranks.Three)
+        };
+
+        [Fact]
+        public void AddCards_Given_SomeCards_Add_Cards()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:22]: 2 cards is the minimalistic scenario.

File: Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+            var expectedCards = _someCards.ToList();
+            hand.AddCards(_someCards);
+            var actualCards = hand.Reveal().ToList();
+
+            Assert.Equal(expectedCards, actualCards);
+        }
+
+        [Fact]
+        public void Reveal_Given_Returns_Empty_List_When_No_Cards_Added()
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:34]: Bad name.
`Reveal_Given_NoCards_Returns_NoCards()`

File: Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs (hunk 9 above)
Comment (Almantask) [reply Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:22]: AddCards_Given_EmptyHand_When_2Cards_Adds_2Cards.

Comment (Almantask) [reply Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:22]: Missing a test to adding cards to a deck with already some (1) cards.

File: Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs (hunk 10 above)
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/HandTests.cs:34]: Should be a theory and test the following:
no cards, 1 card, 2 cards.

File: Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+using BootCamp.Chapter.Gambling;
+using Xunit;
+
+namespace BootCamp.Chapter.Tests.GameComponents
+{
+    public class CardsTests
+    {
+        [Theory]
+        [MemberData(nameof(GetCardAndName))]
+        public void ToString_Returns_Expected_String(Card card, string expectedName)
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs:13]: ..Returns_CardName

Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs:13]: I think this can be a fact and just check if one card works. There is nothing different in all other cards.

File: Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using BootCamp.Chapter.Gambling;
+using Xunit;
+
+namespace BootCamp.Chapter.Tests.GameComponents
+{
+    public class CardsTests
```
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/CardsTests.cs:9]: Should also test ctor.

File: Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs (hunk 6 above)
Comment (Almantask) [inline Tests/BootCamp.Chapter.Tests/GameComponents/DeckTests.cs:26]: Missing a test to draw last card, middle card.

**Review by Almantask — CHANGES_REQUESTED** [review]
Test case names are okay, but could be improved a lot.
Missing edge case test cases.

**Comment by Almantask** [comment]
Also missing ctor tests for Deck

**Comment by Almantask** [comment]
Expired
